### PR TITLE
Render "sizes" attribute for amp-img

### LINKED ISSUE #7500: Blurry images due to small images in `srcset`
## Bug Description

There were a few reports where images were blurry due to how `srcset` and `sizes` are applied on images. The browser will unexpectedly pick a low-res thumbnail instead of a more reasonable high-quality thumbnail.

## Expected Behaviour

No blurry images

## Steps to Reproduce

**Note:** depends on the viewport.

On this story: https://google-webstories.rt.gw/web-stories/batch-1-story-8-home-and-garden/

With the following markup on the first page:

```html
<amp-img
        layout="fill"
         src="https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1.jpg"
        alt="batch1_story8_homeandgarden_screen1_img1"
        srcset="
        https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1.jpg 977w,
        https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1-760x1024.jpg 750w,
        https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1-223x300.jpg 223w"
        class="i-amphtml-layout-fill i-amphtml-layout-size-defined"
        i-amphtml-layout="fill">
</amp-img>
```

AMP runtime will turn it into:

**Note:** `1540px` below is dynamic depending on the viewport.

```html
<amp-img
    layout="fill"
    src="https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1.jpg"
    alt="batch1_story8_homeandgarden_screen1_img1"
    srcset="
    https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1.jpg 977w,
    https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1-760x1024.jpg 750w,
    https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1-223x300.jpg 223w"
    class="i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-element i-amphtml-built i-amphtml-layout"
    i-amphtml-layout="fill"
    style="--loader-delay-offset:84ms !important;">
    <img
        decoding="async"
        sizes="(max-width: 1540px) 72px, 100vw"
        alt="batch1_story8_homeandgarden_screen1_img1"
        srcset="
        https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1.jpg 977w,
        https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1-760x1024.jpg 750w,
        https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1-223x300.jpg 223w"
        src="https://google-webstories.rt.gw/wp-content/uploads/2021/04/batch1_story8_homeandgarden_screen1_img1-1.jpg"
        class="i-amphtml-fill-content i-amphtml-replaced-content"
    />
</amp-img>
```

## Screenshots

See how the browser chooses the 223w version, even though it's smaller than the actual size on the page

Screenshot 2021-05-12 at 18 56 14

Screenshot 2021-05-12 at 18 56 09

Removing the `position: absolute` on the image somewhat improves the experience. Seems like this trips up the browser's `srcset` picking

Screenshot 2021-05-12 at 19 04 04
Screenshot 2021-05-12 at 19 03 57

## Additional Context

- Plugin Version: 1.6.0
- WordPress Version: 5.7
- Operating System: macOS
- Browser: Chrome

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by dreamofabear**
## Context

`sizes` should always accompany `srcset` attribute for responsive image selection. We were relying on AMP's autogeneration of `sizes` which sometimes yields bad results (#7500).

## Summary

Render `sizes` attribute for image output (amp-img) for desktop and mobile viewing modes.

## Relevant Technical Choices

Use the image element width to determine its size in the story.

## To-do

The [`ApplyCommonAttributes`](https://github.com/ampproject/amp-toolbox/blob/main/packages/optimizer/lib/transformers/ApplyCommonAttributes.js) AMP optimizer transform (part of the `ServerSideRendering` transform) moves `sizes` to CSS media queries, which means the `amp-img[sizes]` autogeneration still happens. I assume this is WAI but needs more testing.

## User-facing changes

`srcset` selection should be more consistent and never result in blurry (low resolution) images.

### QA

This PR can be tested by following these steps:

1. Insert images with `srcset` into a story (e.g. Unfold images).
2. Preview the story across a variety of screen widths.
3. Make sure images look reasonable (not blurry).

### UAT

- [x] UAT should use the same steps as above.

### Does this PR have a security-related impact?

No.

### Does this PR change what data or activity we track or use?

No.

### Does this PR have a legal-related impact?

No.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7500

### CODE REVIEW

**Comment by swissspidy** [comment]
Unfortunately now with this PR images are stretched on the frontend.

Before:

Before

After:

After

For the landscape picture, the browser even selected the lower-res 150x100px version instead of the 300x200px version as before. Its markup:

```html
<amp-img layout="fill" src="https://web-stories.local/wp-content/uploads/2021/05/norway-1.webp" alt="norway" srcset="https://web-stories.local/wp-content/uploads/2021/05/norway-1.webp 320w,https://web-stories.local/wp-content/uploads/2021/05/norway-1-300x201.webp 300w,https://web-stories.local/wp-content/uploads/2021/05/norway-1-150x100.webp 150w" sizes="(min-width: 1024px) 17vh, 39vw"></amp-img>
```

> The ApplyCommonAttributes AMP optimizer transform (part of the ServerSideRendering transform) moves sizes to CSS media queries, which means the amp-img[sizes] autogeneration still happens. I assume this is WAI but needs more testing.

The right PHP code in question would be here:

https://github.com/ampproject/amp-toolbox-php/blob/9fe302c68ea917d1e3531b06782c6b9919b6f5c8/src/Optimizer/Transformer/ServerSideRendering.php

From what I can tell this is WAI, as the issue above also happens without this.

File: assets/src/edit-story/elements/image/output.js [hunk 1]
```diff
@@ -42,6 +44,15 @@ function ImageOutput({ element, box }) {
...
     props.srcSet = srcSet;
+
+    // If `srcset` exists but `sizes` doesn't, amp-img will generate a sizes attribute
+    // with best-guess values that can result in poor image selection.
+    const imageWidthPercent = element.width / PAGE_WIDTH;
+    const mobileWidth = Math.round(imageWidthPercent * 100) + 'vw';
+    // Width of a story page in desktop mode is 45vh.
+    const desktopWidth = Math.round(imageWidthPercent * 45) + 'vh';
+    // 1024px is the minimum width for STAMP desktop mode.
+    props.sizes = `(min-width: 1024px) ${desktopWidth}, ${mobileWidth}`;
```
Comment (spacedmonkey) [inline assets/src/edit-story/elements/image/output.js:55]: This is very different from how WordPress uses the size attributes. Take this example. 

```html
<img
loading="lazy"
width="1024"
height="640"
src="http://localhost:8899/wp-content/uploads/2021/06/download-1024x640.jpeg"
alt=""
class="wp-image-1021"
srcset="http://localhost:8899/wp-content/uploads/2021/06/download-1024x640.jpeg 1024w, http://localhost:8899/wp-content/uploads/2021/06/download-300x188.jpeg 300w, http://localhost:8899/wp-content/uploads/2021/06/download-768x480.jpeg 768w, http://localhost:8899/wp-content/uploads/2021/06/download-1536x960.jpeg 1536w, http://localhost:8899/wp-content/uploads/2021/06/download-1200x750.jpeg 1200w, http://localhost:8899/wp-content/uploads/2021/06/download-150x94.jpeg 150w, http://localhost:8899/wp-content/uploads/2021/06/download.jpeg 1920w"
sizes="(max-width: 1024px) 100vw, 1024px"
>
```

It uses a max-width that is the same size as the source image. 

Can you explain the reasoning being this size.

**Comment by dreamofabear** [comment]
Ah, the stretching is caused by the transform extracting `sizes` attribute to CSS `width` plus some cleverness in STAMP code that [rewrites CSS viewport units](https://github.com/ampproject/amphtml/blob/e0285cd4cdd6340466e923391fe32f7f4f47dfb2/extensions/amp-story/1.0/amp-story.js#L590-L596) into `var(--story-page-vw)`, which causes the width to be wrong.

**Comment by dreamofabear** [comment]
OK, there's an interesting interaction here:

1. AMP also supports `sizes` for non-images, so a `sizes` attribute [results in an inline style being added at runtime](https://amp.dev/documentation/guides-and-tutorials/learn/common_attributes/?format=websites#sizes). Confusingly, this is conflated with the `amp-img[sizes]` attribute.
2. Dynamic inline styles blocks SSR boilerplate removal, so the Optimizer's SSR transform extracts the `sizes` attribute into the `<style amp-custom>` tag.
3. The `sizes` attribute is now missing on the amp-img and gets auto-generated, lol.

Up til now, this would actually be fine for (e.g. non-STAMPs), but:

4. amp-story rewrites viewport units in CSS e.g. into `var(--story-page-vw)`.

Now since (3) happens before (4), the auto-generated `sizes` is too large (`1vw` is greater than `1 * var(--story-page-vw)` on desktop) which can result in poor image srcset selection.

The fix is we can add `disable-inline-width` attribute to the amp-img which short-circuits the dance by disabling (1).

File: assets/src/edit-story/elements/image/output.js (hunk 1 above)
Comment (dreamofabear) [reply assets/src/edit-story/elements/image/output.js:55]: Sure, `min-width: 1024px` comes from the STAMP runtime that toggles the desktop viewing mode: https://github.com/ampproject/amphtml/blob/f8f7c7dbec61cec4602e47b00dc883f9bc68ded1/extensions/amp-story/1.0/amp-story.js#L135

That means at screen widths >= 1024px, the browser should use `desktopWidth` as the width of the image for `srcset` selection.

**Comment by dreamofabear** [comment]
Strangely, I'm not able to reproduce the original bug in an imported copy of the "Create Your Indoor Garden Oasis" template (even on the same browser size), but I did some testing and the image selection looks reasonable with this fix. 

Note that the browser image selection also takes `window.devicePixelRatio` into account, so larger images than expected may be selected depending on the test device (I'm using a MBP).

File: includes/KSES.php [hunk 2]
```diff
@@ -541,6 +541,7 @@ public function filter_kses_allowed_html( $allowed_tags ) {
 				'object-fit'                => true,
 				'object-position'           => true,
 				'placeholder'               => true,
+				'sizes'                     => true,
```
Comment (swissspidy) [inline includes/KSES.php:545]: Now need to add `disable-inline-width` here as well.

We could add a new test to `\Google\Web_Stories\Tests\KSES::data_test_filter_kses_allowed_html()` for this new image output

**Comment by swissspidy** [comment]
@googlebot I consent

**Comment by swissspidy** [comment]
Nice writeup, very insightful! TIL about `disable-inline-width` in AMP 💡 

I did some testing with the current solution and it seems to work well so far. The browser's selecting higher-res versions of images and they aren't stretched either.

File: includes/KSES.php (hunk 2 above)
Comment (swissspidy) [reply includes/KSES.php:545]: Added now.

**Review by swissspidy — APPROVED** [review]

**Review by spacedmonkey — APPROVED** [review]
👍

**Pull request merged**